### PR TITLE
gh-109413: Run mypy on `libregrtest` in CI

### PR DESCRIPTION
**Pull request by AlexWaygood**
- Fix remaining mypy errors
- Add a CI check to stop typing issues from regressing in the future

* Issue: gh-109413

### CODE REVIEW

File: Lib/test/libregrtest/results.py [hunk 1]
```diff
@@ -34,7 +34,7 @@ def __init__(self):
         self.test_times: list[tuple[float, TestName]] = []
         self.stats = TestStats()
         # used by --junit-xml
-        self.testsuite_xml: list[str] = []
+        self.testsuite_xml: list = []
```
Comment (AlexWaygood) [inline Lib/test/libregrtest/results.py:37]: This annotation is incorrect; leaving it as it is causes mypy to issue 5 or so errors in various places. We should correct the annotation, ideally, but for now making it more vague gets rid of the errors from mypy. I prefer to defer fixing it properly to a future PR, for now.

File: Lib/test/libregrtest/main.py [hunk 2]
```diff
@@ -371,7 +373,8 @@ def finalize_tests(self, coverage: trace.CoverageResults | None) -> None:
                 os.unlink(self.next_single_filename)
 
         if coverage is not None:
-            coverage.write_results(show_missing=True, summary=True,
+            # uses a new-in-py313 keyword argument that mypy doesn't know about yet:
```
Comment (hugovk) [inline Lib/test/libregrtest/main.py]:
```suggestion
            # uses a new-in-Python 3.13 keyword argument that mypy doesn't know about yet:
```

File: Lib/test/libregrtest/utils.py [hunk 3]
```diff
@@ -580,9 +580,10 @@ def display_header(use_resources: tuple[str, ...],
     print("== Python build:", ' '.join(get_build_info()))
     print("== cwd:", os.getcwd())
 
-    cpu_count = os.cpu_count()
+    cpu_count: object = os.cpu_count()
     if cpu_count:
-        process_cpu_count = os.process_cpu_count()
+        # The function is new in py313; mypy doesn't know about it yet:
```
Comment (hugovk) [inline Lib/test/libregrtest/utils.py]:
```suggestion
        # The function is new in Python 3.13; mypy doesn't know about it yet:
```

**Review by hugovk — APPROVED** [review]
For the new stuff that mypy doesn't know about, can they be ignored via mypy.ini?

If so, that would be cleaner, no need to temporarily add the extra comments in py files.

Otherwise looks good.

**Comment by AlexWaygood** [comment]
> For the new stuff that mypy doesn't know about, can they be ignored via mypy.ini?
> 
> If so, that would be cleaner, no need to temporarily add the extra comments in py files.

I don't think so, unfortunately :(

Mypy's config supports ignoring all errors in a file, or ignoring all errors _of a certain category_ in a file. But I don't think there's a config-file option to ignore all errors due to specific functions or keywords not existing, unfortunately :/

(And, to be fair, that comes up more in CPython than elsewhere! We're literally writing the stdlib here, after all 😄)

**Review by vstinner — APPROVED** [review]
LGTM. These changes seem reasonable to me.

**Comment by AlexWaygood** [comment]
Thanks both!

**Comment by AlexWaygood** [comment]
> [GH-112605](https://github.com/python/cpython/pull/112605) is a backport of this pull request to the [3.12 branch](https://github.com/python/cpython/tree/3.12).

(Not actually backporting the change to the `.yml` file; we won't actually be running mypy on the 3.12 branch in CI. This just backports some of the changes to the `.py` files, to make future backports easier.)

**Comment by vstinner** [comment]
Thanks for enabling mypy on libregrtest :-)

**Pull request merged**